File: fixtures/scale/kotlin/medium/src/calc/parser/Precedence.kt
package calc.parser

import calc.lexer.TokenType

const val LOWEST = 0

fun precedenceOf(type: TokenType): Int = when (type) {
    TokenType.PLUS, TokenType.MINUS -> 10
    TokenType.STAR, TokenType.SLASH, TokenType.PERCENT -> 20
    TokenType.CARET -> 30
    else -> LOWEST
}

fun rightAssociative(type: TokenType): Boolean = type == TokenType.CARET

/** The minimum precedence for the right-hand side of the given operator. */
fun nextLevelFor(type: TokenType): Int =
    if (rightAssociative(type)) precedenceOf(type) - 1 else precedenceOf(type)

File: fixtures/scale/dart/medium/lib/parser/precedence.dart
import '../lexer/token_type.dart';

const lowest = 0;

int precedenceOf(TokenType type) {
  switch (type) {
    case TokenType.plus:
    case TokenType.minus:
      return 10;
    case TokenType.star:
    case TokenType.slash:
    case TokenType.percent:
      return 20;
    case TokenType.caret:
      return 30;
    default:
      return lowest;
  }
}

bool rightAssociative(TokenType type) => type == TokenType.caret;

/// The minimum precedence for the right-hand side of the given operator.
int nextLevelFor(TokenType type) =>
    rightAssociative(type) ? precedenceOf(type) - 1 : precedenceOf(type);
